### PR TITLE
10139 - Difference in expiry date between Namex and client results

### PR DESCRIPTION
**Pull request by andrepestana-aot**
*Issue #:* /bcgov/entity#10139

*Description of changes:*
- set expiration_date time to 11:59pm
- added some tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namex license (Apache 2.0).

### CODE REVIEW

File: api/namex/services/name_request/abstract_name_request.py
Comment (severinbeauvais) [inline api/namex/services/name_request/abstract_name_request.py:157]: I'm concerned that "start" is in one timezone (eg, Daylight time) but the target date is in another timezone (eg, Standard time). Adding a fixed number of days won't fix the timezone offset.

Or... does the Alchemy client correctly convert and save the date in the db ???

File: api/namex/models/request.py
Comment (severinbeauvais) [inline api/namex/models/request.py:136]: Yes, as decided by Linda on Nov 24, that we would only look at the state.

The state is set by a daily job that we will now run right after 11:59 pm, so that NRs that expire at that time will have their new state without delay. (Currently there is a 7h delay between the expiry time and the state change, because the job is run at 7am.)

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [inline api/tests/python/services/name_request/test_abstract_name_request.py:37]: This should specifically look for a 8h offset during Standard time or a 7h offset during Daylight time. This can be a test parameter using known clock change values.

Ref: https://in-the-sky.org/timezone.php

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [inline api/tests/python/services/name_request/test_abstract_name_request.py:26]: The hour and minute should be set to test edge cases.

The current values will work even if the timezone offset is incorrect (7h or 8h) because 11-7=4 is the same day as 11-8=3. (And vice versa for 11+7=18 and 11+8=19).

Comment (severinbeauvais) [reply api/tests/python/services/name_request/test_abstract_name_request.py:26]: In other words, these should verify that a start date before the clock change should result in an expiry date with the correct time offset after the clock change.

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (andrepestana-aot) [reply api/tests/python/services/name_request/test_abstract_name_request.py:37]: yes, I know but we should keep track of the dates that it changes. The python libraries do their jobs and the date looks to be calculated correctly at this point.

**Comment by severinbeauvais** [comment]
Please update bcgov/namex/api/namex/VERSION.py.

File: api/namex/services/name_request/abstract_name_request.py
Comment (severinbeauvais) [reply api/namex/services/name_request/abstract_name_request.py]: outdated comment - closing

File: api/namex/services/name_request/name_request.py
Comment (severinbeauvais) [inline api/namex/services/name_request/name_request.py:149]: Does this need the same fix as above? I'm concerned about the possible assignment of a UTC date.

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [inline api/tests/python/services/name_request/test_abstract_name_request.py:45]: Is the purpose of this to verify the UTC offset in expiry_date? It this redundant, since you're checking the hour and minute in the next 2 lines?

If you think this is needed, I'd prefer to see the expected result as a test parameter, because checking for either -8 or -7 will not catch an off-by-one error.

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [reply api/tests/python/services/name_request/test_abstract_name_request.py]: outdated comment - closing

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [inline api/tests/python/services/name_request/test_abstract_name_request.py:32]: Please change to "at 11:59 pm Pacific". (This is so confusing, I just want to be super clear.)

File: api/namex/services/name_request/abstract_name_request.py
Comment (severinbeauvais) [inline api/namex/services/name_request/abstract_name_request.py:154]: If there is no calling function that sets expiry_hour/min, I would be prefer these parameters to be removed -- it may prevent a future logic bug.

(Same with timezone: if it is always UTC, and should always be UTC, remove the param.)

Comment (severinbeauvais) [reply api/namex/services/name_request/abstract_name_request.py:154]: corrected my comment above ^^^

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [inline api/tests/python/services/name_request/test_abstract_name_request.py:17]: I think it would be better to test edge cases instead of 14:11:00 and 11:11:00, since these won't catch off-by-7 or off-by-8 errors.

In fact, I don't understand why lines 16 and 17 don't resolve to the same date ???

File: api/namex/services/name_request/abstract_name_request.py
Comment (severinbeauvais) [reply api/namex/services/name_request/abstract_name_request.py]: fixed

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (andrepestana-aot) [reply api/tests/python/services/name_request/test_abstract_name_request.py]: >In fact, I don't understand why lines 16 and 17 don't resolve to the same date ???

Yes, you're right. As I wasn't checking any dates the test was ok but yes, there was an issue because I was adding days with UTC. I fixed that and added tests to check the edge cases. THANKS!

File: api/namex/services/name_request/abstract_name_request.py
Comment (severinbeauvais) [inline api/namex/services/name_request/abstract_name_request.py:155]: Please change to "at 11:59 pm Pacific" here also :)

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [reply api/tests/python/services/name_request/test_abstract_name_request.py]: The new tests look great !!!

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [reply api/tests/python/services/name_request/test_abstract_name_request.py]: Thanks. It's actually more readable / easier to understand and verify now, too!

File: api/namex/services/name_request/abstract_name_request.py
Comment (severinbeauvais) [inline api/namex/services/name_request/abstract_name_request.py:170]: This might now be "PST" (Pacific Standard Time = winter time). It might be PDT.
I suggest you just change this to "Pacific".
Also lines above.

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (severinbeauvais) [reply api/tests/python/services/name_request/test_abstract_name_request.py]: fixed, thanks

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (andrepestana-aot) [reply api/tests/python/services/name_request/test_abstract_name_request.py]: I changed it to specific offset.

File: api/namex/services/name_request/name_request.py
Comment (andrepestana-aot) [reply api/namex/services/name_request/name_request.py:149]: I don't think we need to change it here. What are you comparing with? (above?)

File: api/tests/python/services/name_request/test_abstract_name_request.py
Comment (andrepestana-aot) [reply api/tests/python/services/name_request/test_abstract_name_request.py]: Added edge cases.

File: api/namex/services/name_request/name_request.py
Comment (severinbeauvais) [reply api/namex/services/name_request/name_request.py:149]: It liks like it can set start_datetime as "now" in UTC. Should that be "now" local?

Comment (andrepestana-aot) [reply api/namex/services/name_request/name_request.py:149]: As it is timezone aware it is ok. If you create 2 datetimes: one with Pacifc time and another in UTC with the same datetime, both when converted to the same timezone should have the same values.

Comment (severinbeauvais) [reply api/namex/services/name_request/name_request.py:149]: OK

**Review by severinbeauvais — APPROVED** [review]
Please make sure you get approval from the real namex experts :)

File: api/namex/resources/payment/payment.py
Comment (severinbeauvais) [inline api/namex/resources/payment/payment.py:391]: That was easy :)

Now how does this get tested?

**Review by thorwolpert — APPROVED** [review]
the tests look ok

**Review by kialj876 — APPROVED** [review]

File: api/namex/resources/payment/payment.py
Comment (andrepestana-aot) [reply api/namex/resources/payment/payment.py:391]: Sorry, I tried to create a test for the request but the tests in namex are really complicate to me and I got blocked by authentication.

**Review by argush3 — APPROVED** [review]

**Pull request merged**
